### PR TITLE
Savon

### PR DESCRIPTION
**Pull request by guigs**
* Add a Heavy metal SOAP client (savon) to make SOAP API calls and parse XML responses
* Access to all Bing Ads API via `SoapyBing::Service`
* Migrate SoapyBing "curated" APIs (`SoapyBing::Ads`, `SoapyBing::Accounts`) to use SoapyBing::Service
* Add country code mapping (since API v11 return target ids instead of country codes)

### CODE REVIEW

File: lib/soapy_bing/ads/campaign_performance_report.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+module SoapyBing
+  class Ads
+    class CampaignPerformanceReport < Report
+      attr_reader :date_range
+
+      def initialize(options)
+        super
+        @date_range = Range.new(
+          Date.parse(options.fetch(:date_start)),
+          Date.parse(options.fetch(:date_end))
```
Comment (RKushnir) [inline lib/soapy_bing/ads/campaign_performance_report.rb]: It seems a bit awkward to me that start/end dates are passed through 2 functions down the stack, are merged into one object here, and then are still used separately in this same class as `date_range.begin` and `date_range.end`. I would either pass them separately all the way, or pass them as a range and decompose in the place of actual use.

Comment (RKushnir) [reply lib/soapy_bing/ads/campaign_performance_report.rb]: Ah, never mind, I missed that this is not new code :)

File: lib/soapy_bing/service.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      def ad_insight         (options = {}); build(:ad_insight,          options); end
+      def bulk               (options = {}); build(:bulk,                options); end
+      def campaign_management(options = {}); build(:campaign_management, options); end
+      def customer_billing   (options = {}); build(:customer_billing,    options); end
+      def customer_management(options = {}); build(:customer_management, options); end
+      def reporting          (options = {}); build(:reporting,           options); end
+      # rubocop:enable all
+
+      def build(bing_ads_service, options)
+        options[:savon_globals] ||= {}
```
Comment (RKushnir) [inline lib/soapy_bing/service.rb]: The caller might not expect that you mutate the input hash.

File: spec/integration/soapy_bing/ads_campaigns_spec.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+RSpec.describe SoapyBing::Ads do
+  subject(:ads) { described_class.new }
+
+  describe '#campaigns', :vcr do
+    let(:entities) { %w[Campaigns] }
+    let(:fixtured_payload) do
+      JSON.parse(File.read(File.join('spec', 'fixtures', 'ads', 'campaigns.json')))
```
Comment (RKushnir) [inline spec/integration/soapy_bing/ads_campaigns_spec.rb]: You could turn this into a method with the fixture path as a parameter:
```ruby
def fixtured_payload(path)
  JSON.parse(File.read(File.join('spec', 'fixtures', "#{path}.json")))
end

fixtured_payload('ads/campaigns')
…
fixtured_payload('ads/campaign_performance_report')
```

File: spec/support/vcr.rb [hunk 4]
```diff
@@ -144,5 +145,20 @@
...
+
+    modified_csv = CSV.generate do |csv|
+      rows.each do |row|
+        csv << row
+      end
+    end
+
+    interaction.response.body = modified_csv
+  end
+
```
Comment (RKushnir) [inline spec/support/vcr.rb:162]: :+1:

**Review by RKushnir — APPROVED** [review]

File: lib/soapy_bing/ads/campaign_performance_report.rb (hunk 1 above)
Comment (guigs) [reply lib/soapy_bing/ads/campaign_performance_report.rb]: I noticed that too but have decided not to change it. But I think it is simple enough that I'll change it.

File: lib/soapy_bing/service.rb (hunk 2 above)
Comment (guigs) [reply lib/soapy_bing/service.rb]: Good catch

**Pull request merged**